### PR TITLE
[Configs] Add foundation for RelayMiner operation configs.

### PR DESCRIPTION
**Pull request by red-0ne**
### Human Summary

* Implement new `RlayMiner` configuration (#246) parser
* Adapt `RelayMiner` server initialization to
  * Source from the config
  * Check if the config matches the on-chain `Supplier`'s data
  * Support multiple proxy server instances 
* Adapt the `RelayMiner` config parser tests
* Adapt the `RelayMiner` unit tests

This PR is based on #257 and deprecates the `RelayMiner` part of it.

## Issue

- #274
- #271

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Run E2E tests locally**: `make test_e2e`
- [ ] **Run E2E tests on DevNet**: Add the `devnet-test-e2e` label to the PR. This is VERY expensive, only do it after all the reviews are complete.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/relayer/config/relayminer_configs_reader.go [hunk 1]
```diff
@@ -4,98 +4,264 @@ import (
...
+	if err != nil {
+		return nil, ErrRelayMinerConfigInvalidNodeUrl.Wrapf(
+			"invalid query node rpc url %s",
+			err.Error(),
+		)
 	}
 
-	if yamlRelayMinerConfig.ProxiedServiceEndpoints == nil {
-		return nil, ErrRelayMinerConfigInvalidServiceEndpoint.Wrapf("proxied service endpoints are required")
+	// Proxies section
```
Comment (Olshansk) [inline pkg/relayer/config/relayminer_configs_reader.go:81]: This might require a refactor, so want to hear your thoughts first. 

Rather than having comments for sections, what if you create small helper functions and have `RelayMinerConfig,` be the receiver? Then the code might look like:

```go
err := relayMinerConfig.HydrateProxies(yamlConfig)
if err != nil {...}
relayMinerConfig.HydrateSuppliers(yamlConfig)
if err != nil {...}
```

File: pkg/relayer/config/relayminer_configs_reader.go [hunk 2]
```diff
@@ -4,98 +4,264 @@ import (
...
+	// At least one proxy is required
+	if len(yamlRelayMinerConfig.Proxies) == 0 {
+		return nil, ErrRelayMinerConfigInvalidProxy.Wrap("no proxies provided")
+	}
+
+	proxies := yamlRelayMinerConfig.Proxies
```
Comment (Olshansk) [inline pkg/relayer/config/relayminer_configs_reader.go:87]: is the copy necessary or can we just iterate over the yaml data directly?

File: pkg/relayer/config/types.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
+package config
+
+import "net/url"
+
+// YAMLRelayMinerConfig is the structure used to unmarshal the RelayMiner config file
+// TODO_DOCUMENT(@red-0ne): Add proper README documentation for yaml config files.
```
Comment (Olshansk) [inline pkg/relayer/config/types.go:6]:
```suggestion
// TODO_DOCUMENT(@red-0ne): Add proper README documentation for yaml config files and update inline comments as well.
```

File: pkg/relayer/config/types.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+type YAMLRelayMinerConfig struct {
+	SigningKeyName string                         `yaml:"signing_key_name"`
+	SmtStorePath   string                         `yaml:"smt_store_path"`
+	Pocket         YAMLRelayMinerPocketConfig     `yaml:"pocket"`
```
Comment (Olshansk) [inline pkg/relayer/config/types.go:10]: Not a fan of the word `Pocket` since it's too generic/overloaded.

How about `Node` or PocketNode`?

File: pkg/relayer/config/types.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	Proxies        []YAMLRelayMinerProxyConfig    `yaml:"proxies"`
+	Suppliers      []YAMLRelayMinerSupplierConfig `yaml:"suppliers"`
+}
+
+// YAMLRelayMinerPocketConfig is the structure used to unmarshal the pocket
+// node URLs section of the RelayMiner config file
+type YAMLRelayMinerPocketConfig struct {
+	QueryNodeGRPCUrl string `yaml:"query_node_grpc_url"`
```
Comment (Olshansk) [inline pkg/relayer/config/types.go:18]: We'll need to add inline comments here (not now) when we add the proper documentation as well.

File: pkg/relayer/config/types.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	// back to the Host header.
+	XForwardedHostLookup bool
+}
+
+// RelayMinerSupplierConfig is the structure resulting from parsing the supplier
+// section of the RelayMiner config file.
+type RelayMinerSupplierConfig struct {
+	// Name is the serviceId corresponding to the current configuration.
+	Name string
+	// Type is the transport protocol used by the supplier, it must match the
```
Comment (Olshansk) [inline pkg/relayer/config/types.go:100]: Why don't we use a `golang iota` for this or a `protobuf enum`?

File: pkg/relayer/config/types.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	// Hosts is a list of hosts advertised on-chain by the supplier, the corresponding
+	// proxy server will accept relay requests for these hosts.
+	Hosts []string
+	// ServiceConfig is the config of the service that relays will be proxied to.
+	ServiceConfig *RelayMinerSupplierServiceConfig
+}
+
+// RelayMinerSupplierServiceConfig is the structure resulting from parsing the supplier
+// service sub-section of the RelayMiner config file.
+// If the supplier is of type "http", it may embed a basic auth structure and
```
Comment (Olshansk) [inline pkg/relayer/config/types.go:112]: I think this is TMI (Too much Information). The reason we have attribute specific comments is to capture this sort of detail.

File: pkg/relayer/proxy/errors.go [hunk 8]
```diff
@@ -15,4 +15,6 @@ var (
 	ErrRelayerProxyInvalidRelayRequest               = sdkerrors.Register(codespace, 7, "invalid relay request")
 	ErrRelayerProxyInvalidRelayResponse              = sdkerrors.Register(codespace, 8, "invalid relay response")
 	ErrRelayerProxyEmptyRelayRequestSignature        = sdkerrors.Register(codespace, 9, "empty relay response signature")
+	ErrRelayerProxyServiceEndpointNotHandled         = sdkerrors.Register(codespace, 10, "service endpoint not handled by relayer proxy")
+	ErrRelayerProxyUnsupportedTransportType          = sdkerrors.Register(codespace, 11, "unsupported proxy transport type")
```
Comment (Olshansk) [inline pkg/relayer/proxy/errors.go:19]: perf my comment above, I really do think we should be using an enum (golang iota) for this.

File: pkg/relayer/proxy/server_builder.go [hunk 9]
```diff
@@ -28,58 +31,68 @@ func (rp *relayerProxy) BuildProvidedServices(ctx context.Context) error {
...
+					endpoint.Url,
 				)
-			default:
-				return ErrRelayerProxyUnsupportedRPCType
 			}
+		}
+	}
+
+	// Build the advertised relay servers map. For each service's endpoint, create the appropriate RelayServer.
+	providedServices := make(map[string]relayer.RelayServer)
```
Comment (Olshansk) [inline pkg/relayer/proxy/server_builder.go:70]: Another opportunity to use smaller helper functions

File: pkg/relayer/proxy/synchronous.go [hunk 10]
```diff
@@ -82,18 +85,44 @@ func (sync *synchronousRPCServer) Stop(ctx context.Context) error {
...
-
 // ServeHTTP listens for incoming relay requests. It implements the respective
 // method of the http.Handler interface. It is called by http.ListenAndServe()
 // when synchronousRPCServer is used as an http.Handler with an http.Server.
 // (see https://pkg.go.dev/net/http#Handler)
 func (sync *synchronousRPCServer) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
 	ctx := request.Context()
 
+	var originHost string
+	// Get the original host from X-Forwarded-Host header if specified in the proxy config.
```
Comment (Olshansk) [inline pkg/relayer/proxy/synchronous.go:96]: Why do we ned to do this?

File: pkg/relayer/proxy/synchronous.go [hunk 11]
```diff
@@ -82,18 +85,44 @@ func (sync *synchronousRPCServer) Stop(ctx context.Context) error {
...
+	}
+
+	if originHost == "" {
+		originHost = request.Host
+	}
+
+	var supplierService *sharedtypes.Service
+	var serviceUrl *url.URL
+
+outer:
```
Comment (Olshansk) [inline pkg/relayer/proxy/synchronous.go:109]: I'm not a fan of using labels in production code. IMO it means we can either:
1. Use a helper that returns a couple variables.
2. Check in the outer loop if there's a value set and break then

**Review by Olshansk — CHANGES_REQUESTED** [review]
@red-0ne Posting a partial review and will finish it later.

Overall, looks good w/ no real blockers but just some cleanup. I think there's an opportunity to use smaller compartmentalized helper functions to help with code maintainability.

Also, I'd rename the PR title to `[Configs] Add foundation for RelayMiner operation configs.`

Screenshot 2023-12-19 at 5 30 05 PM

File: pkg/relayer/cmd/cmd.go [hunk 12]
```diff
@@ -63,9 +64,11 @@ for such operations.`,
 	cmd.Flags().StringVar(&flagRelayMinerConfig, "config", "", "The path to the relayminer config file")
 
 	// Cosmos flags
+	// TODO_TECHDEBT(#256): Remove unneeded cosmos flags.
 	cmd.Flags().String(cosmosflags.FlagKeyringBackend, "", "Select keyring's backend (os|file|kwallet|pass|test)")
-	cmd.Flags().
-		StringVar(&flagCosmosNodeURL, cosmosflags.FlagNode, omittedDefaultFlagValue, "Register the default Cosmos node flag, which is needed to initialize the Cosmos query and tx contexts correctly. It can be used to override the `QueryNodeUrl` and `NetworkNodeUrl` fields in the config file if specified.")
+	cmd.Flags().StringVar(&flagNodeRPCUrl, cosmosflags.FlagNode, omittedDefaultFlagValue, "Register the default Cosmos node flag, which is needed to initialize the Cosmos query and tx contexts correctly. It can be used to override the `QueryNodeUrl` and `NetworkNodeUrl` fields in the config file if specified.")
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply pkg/relayer/cmd/cmd.go:69]: Renamed to `txNodeGRPCUrl`

File: pkg/relayer/config/proxy_http_config_parser.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
+package config
+
+import (
+	"fmt"
+	"net/url"
+)
+
+// parseRelayMinerConfigs populates the proxy fields of the target structure that
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply pkg/relayer/config/proxy_http_config_parser.go:8]: Comments updated.

This function is actually handling the proxy config when its type is `http`, where `Host` needs to be a valid `http` host (eg. not a UNIX socket`.

In the future, when we support `https` for example, a similar `https` dedicated function would look for `tls` configs and knows how to parse them.

As to the `tcp://` scheme, it's only to have a valid url to parse, and make sure its host is valid. Added a comment and prepended a `http` scheme instead.

File: pkg/relayer/config/relayminer_configs_reader.go [hunk 14]
```diff
@@ -4,98 +4,264 @@ import (
...
+	// SigningKeyName is required
+	if yamlRelayMinerConfig.SigningKeyName == "" {
+		return nil, ErrRelayMinerConfigInvalidSigningKeyName
 	}
 
-	// Parse the query node URL
-	queryNodeUrl, err := url.Parse(yamlRelayMinerConfig.QueryNodeUrl)
-	if err != nil {
-		return nil, ErrRelayMinerConfigInvalidQueryNodeUrl.Wrapf("%s", err)
+	// SmtStorePath is required
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply pkg/relayer/config/relayminer_configs_reader.go:32]: Not sure this is where we should indicate if it's relative or absolute path, unless we check the directory existence which I'm not against.

File: pkg/relayer/config/relayminer_configs_reader.go [hunk 15]
```diff
@@ -4,98 +4,264 @@ import (
...
+			err.Error(),
+		)
 	}
 
-	// Parse the websocket URL of the Pocket Node to connect to for subscribing to on-chain events.
-	pocketNodeWebsocketUrl := sdk.HostToWebsocketURL(queryNodeUrl.Host)
+	// If the query node grpc url is empty, use the tx node grpc url
+	if pocket.QueryNodeGRPCUrl == "" {
+		relayMinerPocketConfig.QueryNodeGRPCUrl = relayMinerPocketConfig.TxNodeGRPCUrl
+	} else {
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply pkg/relayer/config/relayminer_configs_reader.go:57]: Not sure what you want to achieve here.

We want to use `TxNodeGRPCUrl` which is already validated above., if `QueryNodeGRPCUrl` is not set.
Otherwise (`QueryNodeGRPCUrl` has its own value) we need to make sure it's a valid URL.

File: pkg/relayer/proxy/errors.go (hunk 8 above)
Comment (red-0ne) [reply pkg/relayer/proxy/errors.go:19]: Do you mean, getting rid of `sdkerrors.Register`? How to not lose the info captured by the `Register` error description?

File: pkg/relayer/proxy/synchronous.go (hunk 10 above)
Comment (red-0ne) [reply pkg/relayer/proxy/synchronous.go:96]: Updated the comments to explain why we have this.

File: testutil/testproxy/relayerproxy.go [hunk 16]
```diff
@@ -209,14 +216,25 @@ func WithDefaultSessionSupplier(
 // MarshalAndSend marshals the request and sends it to the provided service
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply testutil/testproxy/relayerproxy.go:218]: #PUC with a TODO and/or explanation

File: pkg/relayer/config/relayminer_configs_reader.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply pkg/relayer/config/relayminer_configs_reader.go]: I thought it was just an easier way to write the code (more readable) but feel free to ignore if that's not the case from your POV.

File: pkg/relayer/proxy/errors.go (hunk 8 above)
Comment (Olshansk) [reply pkg/relayer/proxy/errors.go:19]: Not for the errors, for the transport type.

Comment (red-0ne) [reply pkg/relayer/proxy/errors.go:19]: Done

File: pkg/relayer/config/proxies_config_hydrator.go [hunk 17]
```diff
@@ -0,0 +1,63 @@
+package config
```
Comment (Olshansk) [inline pkg/relayer/config/proxies_config_hydrator.go:1]: All of these hydrators are SO easy to read and understand!

File: pkg/relayer/config/relayminer_configs_reader_test.go [hunk 18]
```diff
@@ -16,7 +16,7 @@ func Test_ParseRelayMinerConfigs(t *testing.T) {
 	tests := []struct {
```
Comment (Olshansk) [inline pkg/relayer/config/relayminer_configs_reader_test.go:16]: I'm going to skip reviewing this file 😅

File: pkg/relayer/config/relayminer_configs_reader.go [hunk 19]
```diff
@@ -1,102 +1,68 @@
...
 
-	relayMinerCMDConfig := &RelayMinerConfig{
-		QueryNodeUrl:            queryNodeUrl,
-		NetworkNodeUrl:          networkNodeUrl,
-		PocketNodeWebsocketUrl:  pocketNodeWebsocketUrl,
-		SigningKeyName:          yamlRelayMinerConfig.SigningKeyName,
-		ProxiedServiceEndpoints: proxiedServiceEndpoints,
-		SmtStorePath:            yamlRelayMinerConfig.SmtStorePath,
+	// Check if a proxy is referencing a host more than once
+	for _, proxyConfig := range relayMinerConfig.Proxies {
```
Comment (Olshansk) [inline pkg/relayer/config/relayminer_configs_reader.go:51]: Can you add a `Validate` function (with the appropriate receiver as the config) so this is just a one liner that returns an error?

File: pkg/relayer/config/types.go [hunk 20]
```diff
@@ -0,0 +1,141 @@
+package config
+
+import "net/url"
+
+type ProxyType int
+
+const (
+	ProxyTypeHTTP ProxyType = iota
```
Comment (Olshansk) [inline pkg/relayer/config/types.go:8]: Please add a TODO on what we want to support in the future

File: pkg/relayer/config/types.go [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// RelayMinerSupplierServiceConfig is the structure resulting from parsing the supplier
+// service sub-section of the RelayMiner config file.
+type RelayMinerSupplierServiceConfig struct {
+	// Url is the URL of the service that relays will be proxied to.
+	Url *url.URL
+	// Authentication is the basic auth structure used to authenticate to the
+	// request being proxied from the current proxy server.
+	// If `RelayMinerSupplierConfig.Type` is `ProxyTypeHTTP` then `Authentication`
```
Comment (Olshansk) [inline pkg/relayer/config/types.go:126]: Whe we go type specific, can you add a `For example`. Ditto below

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+  tx_node_grpc_url: tcp://sequencer-poktroll-sequencer:36658
+
+# Proxies are endpoints that expose different suppliers to the internet.
+proxies:
+    # Name of the proxy. It will be used to reference in a supplier. Must be unique.
+    # Required.
+    # TODO_CONSIDERATION: if we enforce DNS compliant names, it can potentially
+    # become handy in the future.
+    # More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names
+  - name: http-example
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:29]: Can you add a `TODO_CONSIDERATION: Rename to supplier_name`

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+  #     key: /path/to/key
+
+# Suppliers are different services that can be offered through RelayMiner.
+# When a supplier is configured to use a proxy and staked appropriately,
+# the relays will start flowing through RelayMiner.
+suppliers:
+    # Name of the supplier offered to the network.
+    # Must be unique.
+    # Required.
+  - name: ethereum
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:57]: Does this nee dto match the on-chain `ServiceId`? If so, #PUC

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+    # Type of how the supplier offers service through the network.
+    # Must match the type of the proxy the supplier is connected to.
+    # Required.
+    type: http
+    # Configuration of the service offered through RelayMiner.
+    service_config:
+      # URL RelayMiner proxies the requests to.
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:64]:
```suggestion
      # URL RelayMiner proxies the requests to.
      # Also known as the data node, or service node in some cases.
```

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+    # Required.
+    proxy_names:
+      - http-example # when the RelayMiner server builder runs.
+  - name: 7b-llm-model
+    type: http
+    service_config:
+      url: http://llama-endpoint
+    hosts:
+      - 7b-llm-model.devnet1.poktroll.com
+      # - 7b-llm-model # <- this part can be added automatically.
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:106]: can or will?

I'm still a bit confused why hosts is a list so #PUC

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+    # request to the correct supplier.
+    # That hostname is what the user should stake the supplier for.
+    # Must be unique within a proxy/proxies it is set up on;
+    # in other words, one proxy can't offer the same hostname more than once.
+    # The `name` of the supplier is automatically added to the hosts section
+    # for potential troubleshooting/debugging purposes
+    # Required.
+    hosts:
+      - ethereum.devnet1.poktroll.com
+      # - ethereum # <- this part is be added automatically.
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:92]: Can you explain WHY? for what purposes / use case?
```suggestion
      # - ethereum # <- this part is added automatically.
```

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+      headers: {}
+
+    # A list of hosts the HTTP service is offering.
+    # When linked to the proxy, that hostname is going to be used to route the
+    # request to the correct supplier.
+    # That hostname is what the user should stake the supplier for.
+    # Must be unique within a proxy/proxies it is set up on;
+    # in other words, one proxy can't offer the same hostname more than once.
+    # The `name` of the supplier is automatically added to the hosts section
+    # for potential troubleshooting/debugging purposes
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:88]: This is a bit wordy and confusing.

When we stake "stake", try to add details on "has the on-chain equivalent when staking .... "

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+      # Optional.
+      headers: {}
+
+    # A list of hosts the HTTP service is offering.
+    # When linked to the proxy, that hostname is going to be used to route the
+    # request to the correct supplier.
+    # That hostname is what the user should stake the supplier for.
+    # Must be unique within a proxy/proxies it is set up on;
+    # in other words, one proxy can't offer the same hostname more than once.
+    # The `name` of the supplier is automatically added to the hosts section
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:87]: Can you move this comment next to the name section and provide more details on why & how it can be used for troubleshooting? E.g. updating /etc/hosts or a custom k8s cluster, etc...

**Review by Olshansk — CHANGES_REQUESTED** [review]
I really like this after the last iteration. 

Can't say I reviewed everything line by line, but it's probably the best codebase for configurations that I've ever seen!

File: localnet/poktrolld/config/appgate_server_config_example.yaml [hunk 29]
```diff
@@ -0,0 +1,12 @@
+# Pocket node URL that exposes CometBFT JSON-RPC API.
+# This can be used by the Cosmos client SDK, event subscriptions, etc...
+query_node_rpc_url: tcp://sequencer-poktroll-sequencer:36657
+# Pocket node URL that exposes the Cosmos gRPC service, dedicated to querying purposes.
+# If unspecified, defaults to `tx_node_grpc_url`.
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply localnet/poktrolld/config/appgate_server_config_example.yaml:5]: Yes, `AppGateServer` don't need a tx node.

File: pkg/relayer/interface.go [hunk 30]
```diff
@@ -64,7 +64,7 @@ type RelayerProxy interface {
 	VerifyRelayRequest(
 		ctx context.Context,
 		relayRequest *servicetypes.RelayRequest,
-		service *sharedtypes.Service,
+		serviceId *sharedtypes.Service,
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply pkg/relayer/interface.go:67]: Good catch!

File: pkg/relayer/config/types.go (hunk 21 above)
Comment (red-0ne) [reply pkg/relayer/config/types.go:126]: Good catch! I confused it with proxy type related fields. This is independent of the proxy type indeed.

File: localnet/poktrolld/config/relayminer_config_full_example.yaml (hunk 22 above)
Comment (red-0ne) [reply localnet/poktrolld/config/relayminer_config_full_example.yaml:29]: It would rather be `proxy_name` since it identifies a `proxy`

File: localnet/poktrolld/config/relayminer_config_full_example.yaml (hunk 25 above)
Comment (red-0ne) [reply localnet/poktrolld/config/relayminer_config_full_example.yaml:106]: This **is** added automatically.

@okdas  may provide more arguments.

The main reasons I see are
* That the operator may want to route the requests coming from these domains to a single proxy
* Migrating to a new domain, where requests may still arrive from the old one 
* Having internal requests coming from non FQDNs because of complex routing and debugging purposes.

File: localnet/poktrolld/config/relayminer_config_full_example.yaml (hunk 28 above)
Comment (red-0ne) [reply localnet/poktrolld/config/relayminer_config_full_example.yaml:87]: Moved to the right section but will let @okdas explain how this is useful. I will then update the comment accordingly.

File: localnet/poktrolld/config/relayminer_config_full_example.yaml (hunk 25 above)
Comment (okdas) [reply localnet/poktrolld/config/relayminer_config_full_example.yaml]: @red-0ne is correct. If we can support multiple, I'd say this is the way. It allows for easier migrations across domains, it gives more flexibility. Kubernetes Ingresses work the same way, heres what ChatGPT has to say about why that is: https://chat.openai.com/share/286502d2-021b-4caa-8c90-eb6138a57c01 - not all points apply to our case (yet)

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 31]
```diff
@@ -21,9 +21,11 @@ pocket_node:
 
 # Proxies are endpoints that expose different suppliers to the internet.
 proxies:
-    # Name of the proxy. It will be used to reference in a supplier. Must be unique.
+    # Name of the proxy. It will be used to reference in a supplier.
+    # Must be unique.
     # Required.
     # TODO_CONSIDERATION: if we enforce DNS compliant names, it can potentially
+    # TODO_CONSIDERATION: Rename to `proxy_name`
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:28]: I personally think we should do it. Wdyt @okdas @red-0ne?

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 32]
```diff
@@ -78,18 +83,25 @@ suppliers:
...
-    # When linked to the proxy, that hostname is going to be used to route the
-    # request to the correct supplier.
-    # That hostname is what the user should stake the supplier for.
-    # Must be unique within a proxy/proxies it is set up on;
-    # in other words, one proxy can't offer the same hostname more than once.
-    # The `name` of the supplier is automatically added to the hosts section
-    # for potential troubleshooting/debugging purposes
+    # A list of hosts the proxy is accepting requests from.
+    # When linked to the proxy, each host is going to be used to lookup the
+    # the Supplier.Service in the Pocket Network.
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:88]:
```suggestion
    # the Supplier.Service in Pocket Network.
```

File: localnet/poktrolld/config/relayminer_config_full_example.yaml [hunk 33]
```diff
@@ -78,18 +83,25 @@ suppliers:
...
+    # Each host in the list must match a Supplier.Service.Endpoint that the Supplier
+    # has advertised on-chain when staking for that Service.
+    # The reasons to have multiple hosts for the same supplier service are:
```
Comment (Olshansk) [inline localnet/poktrolld/config/relayminer_config_full_example.yaml:91]: Same comment as in #288

Screenshot 2023-12-22 at 4 21 03 PM

File: localnet/poktrolld/config/relayminer_config_full_example.yaml (hunk 25 above)
Comment (Olshansk) [reply localnet/poktrolld/config/relayminer_config_full_example.yaml]: @red-0ne I know you already documneted a subset of these, but I think there's opportunity to add even a few more of the points here. Make sure to add the appropriate qualifiers (see my comment)

**Review by Olshansk — CHANGES_REQUESTED** [review]
Will this require additional changes if we go with `proxy_name` and `service_id` instead of `name` or would we want to do this in a followup PR?

File: localnet/poktrolld/config/relayminer_config_full_example.yaml (hunk 25 above)
Comment (red-0ne) [reply localnet/poktrolld/config/relayminer_config_full_example.yaml]: Updated on the `ethereum` example entry

**Review by Olshansk — APPROVED** [review]
Thanks for the multiple iterations on this!

File: localnet/kubernetes/values-relayminer.yaml [hunk 34]
```diff
@@ -1,3 +1,20 @@
...
+      host: 0.0.0.0:8545
+  suppliers:
+    - service_id: anvil
+      type: http
+      service_config:
+        url: http://anvil:8547/
+      proxy_names:
+        - http-proxy
+      hosts:
+        - tcp://relayminers:8545
```
Comment (okdas) [inline localnet/kubernetes/values-relayminer.yaml:20]: Should this be just `relayminers`?

Comment (red-0ne) [reply localnet/kubernetes/values-relayminer.yaml:20]: Since we don't have any proxy/load banalncer yet, the port needs to be added here. Also the `tcp://` scheme is here for url validation, it is then stripped to only take the host part.

**Review by Olshansk — APPROVED** [review]
Let's frikkin 🚢 this thing!
